### PR TITLE
Add `os_disk_type` to `kubernetes_cluster_node_pool`

### LINKED ISSUE #8657: aks: support for ephemeral disks on default & custom node pools
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Add an ability to create node pools with ephemeral disks

### New or Affected Resource(s)

* azurerm_kubernetes_cluster
* azurerm_kubernetes_cluster_node_pool

### Potential Terraform Configuration
```hcl
    default_node_pool {
        xxx
        os_disk_type = "Ephemeral" # or "Managed"
        xxx
    }
```

### References
https://github.com/Azure/AKS/issues/1370
https://github.com/Azure/azure-sdk-for-go/blob/master/services/containerservice/mgmt/2020-09-01/containerservice/models.go#L1379
https://docs.microsoft.com/en-us/rest/api/aks/agentpools/createorupdate#osdisktype

### PR DESCRIPTION
**Pull request by favoretti**
Fixes #8657

### CODE REVIEW

File: azurerm/internal/services/containers/kubernetes_cluster_node_pool_data_source.go [hunk 1]
```diff
@@ -113,6 +113,11 @@ func dataSourceKubernetesClusterNodePool() *schema.Resource {
 				Computed: true,
 			},
 
+			"os_disk_type_gb": {
```
Comment (tombuildsstuff) [inline azurerm/internal/services/containers/kubernetes_cluster_node_pool_data_source.go:116]:
```suggestion
			"os_disk_type": {
```

File: azurerm/internal/services/containers/kubernetes_nodepool.go [hunk 2]
```diff
@@ -121,6 +121,16 @@ func SchemaDefaultNodePool() *schema.Schema {
...
 				},
 
+				"os_disk_type": {
+					Type:     schema.TypeString,
+					Optional: true,
+					ForceNew: true,
+					ValidateFunc: validation.StringInSlice([]string{
+						"Ephemeral",
+						"Managed",
+					}, true),
```
Comment (tombuildsstuff) [inline azurerm/internal/services/containers/kubernetes_nodepool.go:131]: can we make this case-sensitive:

```suggestion
					}, false),
```

File: azurerm/internal/services/containers/kubernetes_nodepool.go [hunk 3]
```diff
@@ -121,6 +121,16 @@ func SchemaDefaultNodePool() *schema.Schema {
 					ValidateFunc: validation.IntAtLeast(1),
 				},
 
+				"os_disk_type": {
+					Type:     schema.TypeString,
+					Optional: true,
+					ForceNew: true,
+					ValidateFunc: validation.StringInSlice([]string{
+						"Ephemeral",
+						"Managed",
```
Comment (tombuildsstuff) [inline azurerm/internal/services/containers/kubernetes_nodepool.go:130]: there's constants for these in the SDK - could we use these:

```
// OSDiskType enumerates the values for os disk type.
type OSDiskType string

const (
	// Ephemeral ...
	Ephemeral OSDiskType = "Ephemeral"
	// Managed ...
	Managed OSDiskType = "Managed"
)
```

```suggestion
						string(containerservice.Ephemeral),
						string(containerservice.Managed),
```

File: azurerm/internal/services/containers/kubernetes_nodepool.go [hunk 4]
```diff
@@ -121,6 +121,16 @@ func SchemaDefaultNodePool() *schema.Schema {
 					ValidateFunc: validation.IntAtLeast(1),
 				},
 
+				"os_disk_type": {
+					Type:     schema.TypeString,
+					Optional: true,
```
Comment (tombuildsstuff) [inline azurerm/internal/services/containers/kubernetes_nodepool.go:126]: presumably this defaults to `Managed` for existing disks, I guess?

File: website/docs/r/kubernetes_cluster_node_pool.html.markdown [hunk 5]
```diff
@@ -95,6 +95,8 @@ The following arguments are supported:
 
 * `os_disk_size_gb` - (Optional) The Agent Operating System disk size in GB. Changing this forces a new resource to be created.
 
+* `os_disk_type` - (Optional) The Agent Operating System disk type. Can be `Managed` or `Ephemeral`. Defaults to `Managed`. Changing this forces a new resource to be created.
```
Comment (tombuildsstuff) [inline website/docs/r/kubernetes_cluster_node_pool.html.markdown:98]: AKS uses [the description for this field](https://docs.microsoft.com/en-us/rest/api/aks/agentpools/createorupdate) `OS disk type to be used for machines in a given agent pool.` for this - but I think we could reword that slightly so it's clearer, so could we make this:

```suggestion
* `os_disk_type` - (Optional) The type of disk which should be used for the Operating System. Possible values are `Ephemeral` and `Managed`. Defaults to `Managed`. Changing this forces a new resource to be created.
```

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @favoretti 

Thanks for this PR :)

Taking a look through this is looking good - if we can fix up the other minor comments then this otherwise LGTM 👍 

Thanks!

File: azurerm/internal/services/containers/kubernetes_nodepool.go (hunk 4 above)
Comment (favoretti) [reply azurerm/internal/services/containers/kubernetes_nodepool.go:126]: Yeah, it defaults to managed, should I add a default here even though it's optional?

Comment (favoretti) [reply azurerm/internal/services/containers/kubernetes_nodepool.go:126]: As discussed, leaving empty.

**Comment by favoretti** [comment]
Running acceptance tests locally now.

**Comment by favoretti** [comment]
```
$ TF_ACC=1 go test -v ./azurerm/internal/services/containers/tests/ -timeout=1000m -run TestAccAzureRMKubernetesClusterNodePool_osDiskType
=== RUN   TestAccAzureRMKubernetesClusterNodePool_osDiskType
=== PAUSE TestAccAzureRMKubernetesClusterNodePool_osDiskType
=== CONT  TestAccAzureRMKubernetesClusterNodePool_osDiskType
--- PASS: TestAccAzureRMKubernetesClusterNodePool_osDiskType (951.34s)
PASS
ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/containers/tests	952.503s
```

**Comment by NillsF** [comment]
@favoretti : Thank you for submitting this PR.
@tombuildsstuff: Will this get merged prior to Nov 19 to ensure it gets included in the next release?

**Review by tombuildsstuff — APPROVED** [review]
LGTM - thanks for pushing those changes @favoretti

**Comment by tombuildsstuff** [comment]
Tests look good 👍

**Comment by djsly** [comment]
FYI, this forces a rebuild for old cluster, unless users reimports / hack their state file with oDiskType: `Managed`

**Comment by favoretti** [comment]
How come? I haven’t seen this effect on 10+ clusters I have.

Can you share a snippet of your cluster + pool definition?

Thanks!

**Comment by djsly** [comment]
had this on our tfplan:
`+ os_disk_type          = "Managed" # forces replacement`
After modifying the state file and adding `"os_disk_type": "Managed"`, it no longer detects any change

**Pull request merged**